File: scripts/global_config.ts
import path from 'path';
import fs from 'fs';
import dotenv from 'dotenv';

// Use __dirname to walk up the directory tree
// In standard CommonJS/Node environments, __dirname is the directory of the current file ('scripts')
// We resolve '..' to get to 'global_agent' root.
export const GLOBAL_AGENT_ROOT = path.resolve(__dirname, '..');

// Load local .env if it exists
const localEnvPath = path.join(GLOBAL_AGENT_ROOT, '.env');
if (fs.existsSync(localEnvPath)) {
    dotenv.config({ path: localEnvPath });
} else {
    console.warn(`⚠️ Warning: Global .env file not found at ${localEnvPath}`);
}

// Check for OS-level override, Local Repo override, then fallback to D:/Assets
const masterFallbackPath = 'D:/Assets/.env.swarm';
const masterEnvPath = process.env.SWARM_ENV_PATH || 
                      (fs.existsSync(localEnvPath) ? localEnvPath : masterFallbackPath);

if (fs.existsSync(masterEnvPath)) {
    dotenv.config({ path: masterEnvPath });
    console.log(`✅ Loaded Master Swarm Environment Cache from ${masterEnvPath}`);
} else {
    console.warn(`⚠️ Warning: Master Swarm Environment Cache not found at ${masterEnvPath}`);
}

// Sibling Repository Resolution
// We assume all repos sit in the same parent directory.
export const GITHUB_ROOT = path.resolve(GLOBAL_AGENT_ROOT, '..');

// Common Ecosystem Repositories
export const ERIKNORRIS_ROOT = path.join(GITHUB_ROOT, 'eriknorris');
export const MECHANISTIC_ROOT = path.join(GITHUB_ROOT, 'mechanistic');
export const MOOTMOAT_ROOT = path.join(GITHUB_ROOT, 'mootmoat');
export const PORTFOLIO_ROOT = path.join(GITHUB_ROOT, 'portfolio');

export function getRepoRoot(repoName: string): string {
    const repos: { [key: string]: string } = {
        'global_agent': GLOBAL_AGENT_ROOT,
        'eriknorris': ERIKNORRIS_ROOT,
        'mechanistic': MECHANISTIC_ROOT,
        'mootmoat': MOOTMOAT_ROOT,
        'portfolio': PORTFOLIO_ROOT
    };
    
    let repoPath = repos[repoName];
    if (!repoPath) {
        repoPath = path.join(GITHUB_ROOT, repoName);
    }
    
    if (!fs.existsSync(repoPath)) {
        throw new Error(`❌ Critical Error: Could not locate sibling repository '${repoName}' at ${repoPath}`);
    }
    
    return repoPath;
}

// Centralized utility paths
export const GLOBAL_INBOX_DIR = path.join(GLOBAL_AGENT_ROOT, '_inbox');

// Ensure directories exist
if (!fs.existsSync(GLOBAL_INBOX_DIR)) {
    fs.mkdirSync(GLOBAL_INBOX_DIR, { recursive: true });
}
const processedDir = path.join(GLOBAL_INBOX_DIR, 'processed');
if (!fs.existsSync(processedDir)) {
    fs.mkdirSync(processedDir, { recursive: true });
}

File: scripts/global_config.py
import os
from pathlib import Path
from dotenv import load_dotenv

# Use Path(__file__) to dynamically walk up the directory tree
GLOBAL_AGENT_ROOT = Path(__file__).resolve().parent.parent
GITHUB_ROOT = GLOBAL_AGENT_ROOT.parent

# Sibling repositories
MECHANISTIC_ROOT = GITHUB_ROOT / "mechanistic"
MOOTMOAT_ROOT = GITHUB_ROOT / "mootmoat"
# Handle portfolio differently since it's typically D:/portfolio based on user environments
PORTFOLIO_ROOT = Path("D:/portfolio") if Path("D:/portfolio").exists() else GITHUB_ROOT / "portfolio"

# Load local .env
local_env = GLOBAL_AGENT_ROOT / ".env"
if local_env.exists():
    load_dotenv(dotenv_path=local_env)
else:
    print(f"⚠️ Warning: Global .env file not found at {local_env}")

# Master Swarm Environment Override
master_fallback = Path("D:/Assets/.env.swarm")
master_env_path = Path(os.getenv("SWARM_ENV_PATH", master_fallback))

if master_env_path.exists():
    load_dotenv(dotenv_path=master_env_path)
    # print(f"✅ Loaded Master Swarm Environment Cache from {master_env_path}")

def get_repo_root(repo_name: str) -> Path:
    repos = {
        'global_agent': GLOBAL_AGENT_ROOT,
        'mechanistic': MECHANISTIC_ROOT,
        'mootmoat': MOOTMOAT_ROOT,
        'portfolio': PORTFOLIO_ROOT
    }
    
    path = repos.get(repo_name)
    if not path:
        path = GITHUB_ROOT / repo_name
        
    if not path.exists():
        raise FileNotFoundError(f"❌ Critical Error: Could not locate sibling repository '{repo_name}' at {path}")
        
    return path

# Grok Engine & Vault configuration
LLAMA_MODEL_PATH = os.getenv("LLAMA_MODEL_PATH", r"D:\Local_Grok_Engine\models\Meta-Llama-3-8B-Instruct.Q4_K_M.gguf")
GROK_VAULT_DIR = os.getenv("GROK_VAULT_DIR", r"H:\Grok_Vector_Vault")
